### ISSUE TITLE
Variables acting as embedded arguments are not checked against regular expression

### ISSUE THREAD

**Issue by MoreFamed** [robotframework/robotframework#4069]
Hi, I came across strange behaviour of RF regarding embedded arguments. I have keyword
```
Perform ${arg:[^ ]+}
    Log To Console  ${arg}
```
In other words, the argument cannot contain any space. Then, `Perform abcd` passes, as it contains no space. `Perform ab cd` fails with _No keyword with name 'Perform ab cd' found_, since it does contain a space. So far so good. But when I set the value with space to a variable and give it to the keyword: `${a}  Set Variable  ab cd` and `Perform ${a}`, the keyword surprisingly passes (and logs _ab cd_). I expect it to fail as if in case when the non-matching value was given to the keyword directly.

**Comment by MoreFamed** [comment]
Hi, and what about converting value of the non-string variable to string and match it against the regexp? And then use its original value if it is accepted. Would it make sense?

**Comment by pekkaklarck** [comment]
Matching non-string values to regular expressions sounds a bit strange, but I guess it could make sense in this context. We'd only check the value if there's a custom regular expression anyway. I'm personally fine doing this either way.

**Comment by pekkaklarck** [comment]
This issue has a big risk of not making it to RF 5.0. It would be nice to have it, but we also want that release out as soon as possible. If you consider this issue important, comment here. If you really want to see it in RF 5.0, consider creating a pull request implementing it.

**Comment by pekkaklarck** [comment]
I had some time for coding on a train ride and decided to look at this myself. I noticed implementation isn't exactly trivial because we don't currently store custom patterns anywhere separately. I got a prototype working now but it needs a little cleanup as well as tests and a doc update.

**Comment by pekkaklarck** [comment]
I initially decided to make it so that if you specify a custom pattern, then the argument must be a string. That turned out to be a bad idea, because we even had acceptance tests with patterns like `\d+` that were called with variables containing integers. That failing makes no sense. I thought about converting non-string values to strings before matching, but that's a bit risky with custom objects. My current thinking is that non-string values should just be always accepted.

**Comment by pekkaklarck** [comment]
This turned out to be quite a bit bigger task than I anticipated. Especially handling the `\` escape character in patterns was pretty complicated. It didn't actually work correctly earlier.

The change is backwards incompatible in two ways:
- Variable values must now match custom regular expressions. This is pretty obvious.
- As mentioned above, handling escapes was buggy earlier. For example, if you wanted to match `c:\\temp\\example` (`\\` needed because `\` is an escape character), the pattern needed to be `${path:c:\\\\temp\\\\.*}` i.e. the `\` needed to be double escaped. The documentation didn't talk anything about double escaping so this was clearly a bug. Handling escapes was changed and now `${path:c:\\temp\\.*}` works.

**Comment by oboehmer** [comment]
The change for this issue will cause us quite some grief in our large keyword library:

We have generally embraced an embedded argument strategy which we feet/felt is easier for our test writer population which generally don't have a programming background.

We have keywords like `execute "${command}"` and `execute "${command}" on device "${device}"`, which normally trigger the ambiguous keyword error (due to robot's greedy search):

```
Test case file contains multiple keywords matching name 'execute "${what}" on device "${device}"':
    execute "${command}"
    execute "${command}" on device "${device}"
```

We have mitigated this issue so far using regexp in the problematic arguments, i.e. we changed the first keyword to `execute "${command:[^"]+}"` 

This has been working fine, but with the recent change, a test case like this stops working once the argument contains a `"` as it doesn't match the regexp:

```
test
    ${what}=   Set Variable   this is a "test"
    execute "${what}"
```

which now triggers `ValueError: Embedded argument 'command' got value 'this is a "test"' that does not match custom pattern '[^"]+'.`

I wonder:
- is there a practical alternative to the `[^"]+` regex to handle the above case (short of deprecating and changing our keywords, which would be **extremely** painful)
- If no alternative exist, can we maybe postpone this change or provide some knob to disable this additional logic?

**Comment by pekkaklarck** [comment]
Thanks for testing the beta and bringing this issue up @oboehmer. Conflicts like you explain are certainly a problem with embedded arguments and using the `[^"]+` pattern is a common way to avoid them. It's actually even recommended in the [User Guide](http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#using-custom-regular-expressions). Breaking this usage is so badly backwards incompatible that it could only be done in a major release like RF 6.0, and even then it would be questionable would the inconvenience match the benefits of more strict argument validation.

The underlying problem here is the keyword conflict. If we could resolve them better, without a need for custom regexps, we could then make the custom regexp validation more strict. There are two approaches for helping with conflicts:
- Come up with some other solution to let users tell the framework which keyword to use if there is a conflict. I just don't have any ideas what this solution could be.
- Enhance Robot itself so that it can resolve conflicts automatically. For example, when using `Execute "x" on device "y"`, it's pretty clear which keyword implementation in the above commit would match it best. I have an idea how to implement that but I'm not sure does it work in all cases. I'll submit a separate issue about that.

I think this is how we should go forward with this issue:
1. Revert changes so that custom regexps aren't validated. We don't need to revert all changes, just disabling the validation is enough, to make it easy to enable it again later if we can.
2. Try to come up with a better conflict resolution system. That requires a separate issue.
3. When we have a better conflict resolution system in place, make the custom regexp validation more strict in the next (major) release.

**Comment by pekkaklarck** [comment]
Although I think these changes now need to be reverted due to backwards incompatibility problems, it is certainly somewhat strange that if you have keyword like

```
*** Keywords ***
Execute "${command:[^"]+}"
    No Operation
```

you cannot use it like `Execute "show "foo""`, but `Execute "${var}"` works regardless what `${var}` contains. It could be argued that if the argument is known to possibly contain `"`, then the argument should be surrounded with some other markers in the keyword name. For example, all these ought to be fine:

```
Execute '${command:[^']+}'
Execute `${command:[^`]+}`
Execute <${command:[^>]+}>
```

Anyway, breaking the old usage in a non-major release isn't a good idea. We can consider this again in RF 6.0 when we hopefully have some better conflict resolution system in place as well.




doing it in a release like 5.1 is certainly not

**Comment by pekkaklarck** [comment]
This functionality was disabled in 28a50a13ff8693112937af3e0bf4c19a9a352348. This issue was accidentally closed because GitHub interpreted the wording in the commit message so, but this is in fact still open and now targeted to RF 5.2. We may want to decide to postpone this further to RF 6.0, though, in which case the old behavior should be loudly deprecated in RF 5.2.

**Comment by pekkaklarck** [comment]
RF 6.0 (previously known as RF 5.1) enhanced conflict resolution significantly (#4454) and silently deprecated variables not matching custom embedded arguments patterns (#4462). We will make that deprecation more loud in RF 6.1 (#4524) and the behavior can then be finally changed in RF 7.0.

**Comment by pekkaklarck** [comment]
The planned loud deprecation of variables not matching patterns (#4524) wasn't done in RF 6.1 and there never was RF 6.2. In RF 7 we now need to decide should we do deprecation (#4524) or should we directly change the behavior (this issue). I'll now initially leave both of these to RF 7 scope, but if we do deprecation then we need to move this to RF 8 scope. If we decide to change the behavior right away, then the other issue can simply be closed.

**Comment by oboehmer** [comment]
Not a surprise, I guess, for me opting to deprecate first. It will allow folks to remove the pattern instead of keywords failing suddenly.

**Comment by pekkaklarck** [comment]
Yeah, let's take the safe route and deprecate this behavior properly first. I'll move this issue in to RF 8 scope.

**Comment by pekkaklarck** [comment]
Interestingly this behavior change affects also using custom patterns case-insensitively. Earlier if you had a keyword like `Select ${animal:cat|dog}`, you could use it as `Select cat` and `Select Cat` without problems. Making the pattern matching strict means that the match much be case-sensitive and as the result there's nowadays a deprecation warning from the latter. I believe matching being case-sensitive is a good thing, but there needs to be a way to make it case-insensitive. Issue #5242 handles that by allowing using inline flags like `(?i)`.
